internal/helm/chart.go

```diff
@@ -57,3 +66,127 @@ func OverwriteChartDefaultValues(chart *helmchart.Chart, data []byte) (bool, err
+// given chartPath. It takes "requirements.yaml" files into account, and is therefore compatible with the
+// chart.APIVersionV1 format.
+func LoadChartMetadata(chartPath string) (meta *helmchart.Metadata, err error) {
+	i, err := os.Stat(chartPath)
+	if err != nil {
+		return nil, err
+	}
+	if i.IsDir() {
+		meta, err = LoadChartMetadataFromDir(chartPath)
+		return
+	}
+	meta, err = LoadChartMetadataFromArchive(chartPath)
+	return
+}
+
+// LoadChartMetadataFromDir loads the chart.Metadata from the "Chart.yaml" file in the directory at the given path.
+// It takes "requirements.yaml" files into account, and is therefore compatible with the chart.APIVersionV1 format.
+func LoadChartMetadataFromDir(dir string) (*helmchart.Metadata, error) {
+	m := new(helmchart.Metadata)
+
+	b, err := os.ReadFile(filepath.Join(dir, chartutil.ChartfileName))
+	if err != nil {
+		return nil, err
+	}
+	err = yaml.Unmarshal(b, m)
+	if err != nil {
+		return nil, fmt.Errorf("cannot load '%s': %w", chartutil.ChartfileName, err)
+	}
+	if m.APIVersion == "" {
+		m.APIVersion = helmchart.APIVersionV1
+	}
+
+	b, err = os.ReadFile(filepath.Join(dir, "requirements.yaml"))
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return nil, err
+	}
+	if len(b) > 0 {
+		if err = yaml.Unmarshal(b, m); err != nil {
+			return nil, fmt.Errorf("cannot load 'requirements.yaml': %w", err)
+		}
+	}
+	return m, nil
+}
+
+// LoadChartMetadataFromArchive loads the chart.Metadata from the "Chart.yaml" file in the archive at the given path.
+// It takes "requirements.yaml" files into account, and is therefore compatible with the chart.APIVersionV1 format.
+func LoadChartMetadataFromArchive(archive string) (*helmchart.Metadata, error) {
+	f, err := os.Open(archive)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	r := bufio.NewReader(f)
+	zr, err := gzip.NewReader(r)
+	if err != nil {
+		return nil, err
+	}
+	tr := tar.NewReader(zr)
+
+	var m *helmchart.Metadata
+	for {
+		hd, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		if hd.FileInfo().IsDir() {
+			// Use this instead of hd.Typeflag because we don't have to do any
+			// inference chasing.
+			continue
+		}
+
+		switch hd.Typeflag {
+		// We don't want to process these extension header files.
+		case tar.TypeXGlobalHeader, tar.TypeXHeader:
+			continue
+		}
+
+		// Archive could contain \ if generated on Windows
+		delimiter := "/"
+		if strings.ContainsRune(hd.Name, '\\') {
+			delimiter = "\\"
+		}
+		parts := strings.Split(hd.Name, delimiter)
+
+		// We are only interested in files in the base directory
+		if len(parts) != 2 {
+			continue
+		}
+
+		// Normalize the path to the / delimiter
+		n := strings.Join(parts[1:], delimiter)
+		n = strings.ReplaceAll(n, delimiter, "/")
+		n = path.Clean(n)
+
+		switch parts[1] {
+		case chartutil.ChartfileName, "requirements.yaml":
+			b, err := io.ReadAll(tr)
```

Comment:
The recommendation from https://github.com/fluxcd/source-controller/issues/470 might apply here? I'm not sure what is a reasonable value for the maximum file size though.

Reply:
Yes, I think it should be a configurable option with a high default first. Didn't take the time yet to factor this in.